File: Runtime/StockFlowTransformer.Maps.vb
'************************************************************************************
' STSimStockFlow: A .NET library for simulating stocks and flows
'
' Copyright © 2007-2017 Apex Resource Management Solution Ltd. (ApexRMS). All rights reserved.
'
'************************************************************************************

Partial Class StockFlowTransformer

    Private m_StockLimitMap As StockLimitMap
    Private m_FlowPathwayMap As FlowPathwayMap
    Private m_FlowMultiplierMap As FlowMultiplierMap
    Private m_FlowSpatialMultiplierMap As FlowSpatialMultiplierMap
    Private m_InitialStockSpatialMap As InitialStockSpatialMap
    Private m_FlowOrderMap As FlowOrderMap

    Private Sub CreateStockLimitMap()

        Debug.Assert(Me.m_StockLimitMap Is Nothing)

        Me.m_StockLimitMap = New StockLimitMap(
            Me.ResultScenario,
            Me.m_StockLimits)

    End Sub

    Private Sub CreateFlowPathwayMap()

        Debug.Assert(Me.m_FlowPathwayMap Is Nothing)
        Me.m_FlowPathwayMap = New FlowPathwayMap(Me.m_FlowPathways)

    End Sub

    Private Sub CreateFlowMultiplierMap()

        Debug.Assert(Me.m_FlowMultiplierMap Is Nothing)

        Me.m_FlowMultiplierMap = New FlowMultiplierMap(
            Me.ResultScenario,
            Me.m_FlowMultipliers,
            Me.m_STSimTransformer.DistributionProvider)

    End Sub

    Private Sub CreateFlowSpatialMultiplierMap()

        Debug.Assert(Me.m_FlowSpatialMultiplierMap Is Nothing)

        Me.m_FlowSpatialMultiplierMap = New FlowSpatialMultiplierMap(
            Me.ResultScenario,
            Me.m_FlowSpatialMultipliers)

    End Sub

    Private Sub CreateInitialStockSpatialMap()

        Debug.Assert(Me.m_InitialStockSpatialMap Is Nothing)
        Me.m_InitialStockSpatialMap = New InitialStockSpatialMap(Me.m_InitialStocksSpatial)

    End Sub

    Private Sub CreateFlowOrderMap()

        Debug.Assert(Me.m_FlowOrderMap Is Nothing)
        Me.m_FlowOrderMap = New FlowOrderMap(Me.m_FlowOrders)

    End Sub

End Class

File: src/Runtime/StockFlowTransformer.Maps.cs
// stsim-stockflow: SyncroSim Add-On Package (to stsim) for integrating stocks and flows into state-and-transition simulation models in ST-Sim.
// Copyright © 2007-2019 Apex Resource Management Solutions Ltd. (ApexRMS). All rights reserved.

using System.Diagnostics;

namespace SyncroSim.STSimStockFlow
{
	internal partial class StockFlowTransformer
	{
		private StockLimitMap m_StockLimitMap;
		private StockTransitionMultiplierMap m_StockTransitionMultiplierMap;
		private FlowPathwayMap m_FlowPathwayMap;
		private InitialStockSpatialMap m_InitialStockSpatialMap;
		private FlowOrderMap m_FlowOrderMap;

		private void CreateStockLimitMap()
		{
			Debug.Assert(this.m_StockLimitMap == null);
			this.m_StockLimitMap = new StockLimitMap(this.ResultScenario, this.m_StockLimits);
		}

		private void CreateStockTransitionMultiplierMap()
		{
			Debug.Assert(this.m_StockTransitionMultiplierMap == null);
			this.m_StockTransitionMultiplierMap = new StockTransitionMultiplierMap(this.ResultScenario, this.m_StockTransitionMultipliers, this.m_STSimTransformer.DistributionProvider);
		}

		private void CreateFlowPathwayMap()
		{
			Debug.Assert(this.m_FlowPathwayMap == null);
			this.m_FlowPathwayMap = new FlowPathwayMap(this.m_FlowPathways);
		}

		private void CreateInitialStockSpatialMap()
		{
			Debug.Assert(this.m_InitialStockSpatialMap == null);
			this.m_InitialStockSpatialMap = new InitialStockSpatialMap(this.m_InitialStocksSpatial);
		}

		private void CreateFlowOrderMap()
		{
			Debug.Assert(this.m_FlowOrderMap == null);
			this.m_FlowOrderMap = new FlowOrderMap(this.m_FlowOrders);
		}

        private void CreateMultiplierTypeMaps()
        {
            foreach (FlowMultiplier tm in this.m_FlowMultipliers)
            {
                FlowMultiplierType mt = this.GetFlowMultiplierType(tm.FlowMultiplierTypeId);
                mt.AddFlowMultiplier(tm);
            }

            foreach (FlowSpatialMultiplier tm in this.m_FlowSpatialMultipliers)
            {
                FlowMultiplierType mt = this.GetFlowMultiplierType(tm.FlowMultiplierTypeId);
                mt.AddFlowSpatialMultiplier(tm);
            }  

            foreach (FlowMultiplierType tmt in this.m_FlowMultiplierTypes)
            {
                tmt.CreateFlowMultiplierMap();
                tmt.CreateSpatialFlowMultiplierMap(); 
            }
        }
    }
}
